File: lib/page/message/http/list_message.dart
// ignore_for_file: constant_identifier_names

import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:comput/util/util.dart';

class Message {
  final String id;
  final String userId;
  final String userName;
  final int type;
  final String content;
  const Message({
    required this.id,
    required this.userId,
    required this.userName,
    required this.type,
    required this.content,
  });

  /// 添加记录
  static Future<dynamic> addMessage(
      {required String userId,
      required String listDataId,
      required int type,
      required String content}) async {
    try {
      final res = await http.post(Uri.parse('${Http.basicUrl}/addMessage'),
          body: {
            "userId": userId,
            "listDataId": listDataId,
            "type": "$type",
            "content": content
          });
      final code = jsonDecode(res.body)['code'];
      return code;
    } catch (e) {
      rethrow;
    }
  }

  /// 撤回记录
  static Future<bool> revokeMessage(
      {required String listDataId, required String messageId}) async {
    try {
      final result = await http.get(Uri.parse(
          '${Http.basicUrl}/message/revoke/$messageId/$listDataId?type=$REVOKE'));
      int code = jsonDecode(result.body)['code'];
      return code == 1000;
    } catch (e) {
      rethrow;
    }
  }

  static const int MESSAGE = 0;
  static const int REVOKE = 1;

  factory Message.fromJson(Map<String, dynamic> json) => Message(
        content: json['content'],
        userId: json['userId'],
        userName: json['userName'],
        type: json['type'] ?? MESSAGE,
        id: json['_id'],
      );
}

class MessageDiff {
  final String id;
  final int type;
  final String content;
  factory MessageDiff.fromJson(Map<String, dynamic> json) => MessageDiff(
      type: json['type'] ?? Message.MESSAGE,
      id: json['_id'],
      content: json['content']);

  MessageDiff({required this.content, required this.id, required this.type});

  static parseString(Message message) {
    return {"id": message.id, "type": message.type, "content": message.content};
  }
}

File: server/router/listMessage.ts
import express from 'express'
import { err, info } from '../util';
import { getUserById } from '../service/user';
import { addListMessage, getListMessageByListDataId, revokeMessage } from '../service/listMessage';
import { getListData } from '../service/listData';

export const listMessageRoute = express.Router();

listMessageRoute.post('/addMessage', async (req, res) => {
  const { userId, listDataId, content, type } = req.body
  if (userId && listDataId && content) {
    try {
      const resUser = await getUserById(userId);
      const [resListData] = await getListData({ _id: listDataId });
      if (resUser && resListData) {
        const result = await addListMessage({
          userId: resUser._id, userName: resUser.name,
          listDataId: resListData._id, content, type: type ?? 0
        })
        return info(res, result);
      } else {
        return err(res, { msg: 'id错误' })
      }
    } catch (e) {
      return err(res, { msg: '查询出错' })
    }
  } else {
    return err(res, { msg: '传值错误' })
  }
})

listMessageRoute.get('/message/revoke/:messageId/:listDataId', async (req, res) => {
  const { messageId, listDataId } = req.params
  const { type } = req.query
  if (messageId && listDataId && type) {
    const result = await revokeMessage(listDataId, messageId, +type);
    if (result.modifiedCount != 0) {
      return info(res, result)
    }
    return err(res, { msg: '更新失败' })
  }
  return err(res, { msg: '缺少必要参数' })
})

listMessageRoute.get('/getListMessage/:listDataId', async (req, res) => {
  const { listDataId } = req.params
  if (listDataId) {
    const result = await getListMessageByListDataId(listDataId)
    info(res, result)
  } else {
    return err(res)
  }
})


